text/0000-rpc-queue.md

```diff
@@ -0,0 +1,86 @@
+
+
+
+![](0000-rpc-queue/RCPQueueInboundHigh.png)
+
+1. The NetworkingManager enumerates through all current transport events (in the case of UNET) and adds any received RPC messages to the current inbound QueueHistoryFrame via the RPCQueueContainer (Network PreUpdate Stage).
+2. Later in the PlayerLoop (Network FixedUpdate stage) the RpcQueueContainer enumerates through all inbound RPCs and as long as they are valid will invoke each RPC, NetworkedObject relative, within the associated NetworkedBehaviour component instance initiated by the sender.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Outbound RPC Data Flow Pipeline
+![](0000-rpc-queue/OutboundDataFlowPipeline.png)
+
+### Inbound RPC Data Flow Pipeline
+![](0000-rpc-queue/InboundDataFlowPipeline.png)
+
+### RPC Queue Classes
+![](0000-rpc-queue/RPCQueueClasses.png)
+
+- **RPCQueueContainer:** Manages inbound and outbound RPC queues.  Both inbound and outbound queues are byte arrays exposed as streams (i.e. currently BitStreams) and contained within a QueueHistoryFrame.  The number of QueueHistoryFrames is defined by calling the Initialize method and passing the maximum number of history frames.  The total number of frames is MaxFrameHistory + 1, where the additional frame is considered the “current frame” (i.e. it will always maintain the exact number of frames in history).
+- **RPCQueueProcessing:** Currently, this class is instantiated by the RPCQueueContainer during its initialization.  It is highly probable that this class will get absorbed into the RPCQueueContainer.
+- **QueueHistoryFrame:** Container class for handling the management of an RPC queue.  One QueueHistoryFrame instance per data flow pipeline (i.e. inbound and outbound yields two QueueHistoryFrames).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This approach does include additional processing overhead and there could be some refactoring depending upon any future transport related work that might occur.  
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+Since there are no current discussions on batching specific packet types together at the transport layer (not recommended) and while Unet provides the ability to switch from immediate to queued mode there were other framework related design factors taken into condiseration (i.e. batched RPCs, invocation at different update stages, etc.) that made this the current best path to take.
+
+
+# Prior art
+[prior-art]: #prior-art
+
+
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
+- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC?
```

Comment:
we want to eliminate memory footprint and overhead we currently have but it's out of scope for this RFC, right?

Reply:
Other than the pre-allocation of the BitStreams for each QueueHistoryFrame, after Matt's updates to the Channel Name (i.e. string to byte) RPCQueueContainer's footprint is close to zero (if not zero).  I still need to run further tests on this, but it seems the only things causing memory allocations is when entities are created.  One can run a deep profile on it and find frames where no entities are being created/destroyed and the GC Alloc is considered zero for RPCQueueContainer and associated classes.